### PR TITLE
Add media type

### PR DESCRIPTION
**Pull request by cholmes**
**Related Issue(s):** #851 #918 

**Proposed Changes:**

1. Added media types: `application/geo+json; profile=stac-item` for Items, `application/json; profile=stac-catalog` for Catalogs, and `application/json; profile=stac-collection` for Collection.

**PR Checklist:**

- [x] This PR is made against the dev branch (all proposed changes except releases should be against dev, not master).
- [x] This PR has **no** breaking changes. (this is arguable, but an additional 'profile' on media types already used isn't supposed to break anything)
- [x] I have added my changes to the [CHANGELOG](https://github.com/radiantearth/stac-spec/blob/dev/CHANGELOG.md) **or** a CHANGELOG entry is not required.
- [x] This PR affects the [STAC API spec](https://github.com/radiantearth/stac-api-spec), and I have opened issue/PR https://github.com/radiantearth/stac-api-spec/issues/100 to track the change.

### CODE REVIEW

File: catalog-spec/catalog-spec.md [hunk 1]
```diff
@@ -71,6 +71,13 @@ The following types are commonly used as `rel` types in the Link Object of a STA
 
 **Note:** A link to at least one `item` or `child` catalog is **REQUIRED**.
 
+## Media Type for STAC Catalogs
+
+A STAC Catalog is a JSON file ([RFC 8259](https://tools.ietf.org/html/rfc8259)), and thus should use the 
+`[application/json](https://tools.ietf.org/html/rfc8259#section-11)` as the [Media Type](https://en.wikipedia.org/wiki/Media_type) 
+(previously known as the MIME Type). STAC Catalogs are recommended to additional specify a `profile` equal to `stac-catalog` to
```
Comment (m-mohr) [inline catalog-spec/catalog-spec.md:78]: In media type terminology `profile` is a parameter. Applies to all three types.

```suggestion
(previously known as the MIME Type). STAC Catalogs are recommended to additionally add the parameter `profile` with the value `stac-catalog` to
```

**Comment by m-mohr** [comment]
There are tables containing media types in the specs, which should add the new media types, too.

**Comment by cholmes** [comment]
Will add them. I had been thinking the lists were asset focused, but looks like they're written generically.

**Comment by cholmes** [comment]
Ok, I think this is fully ready for review. @m-mohr - I reworked the media type section a bit, adding the stac types. It was also focused on 'assets', and under the asset section, which was a bit weird since catalog and collection linked to it directly in their link object descriptions, which didn't make much sense. So now it's pulled out to be a general media type section. I think it reads a bit better, but happy for any feedback / tweaks. Also updated all of them with your language on the parameter - thanks.

File: item-spec/item-spec.md [hunk 2]
```diff
@@ -252,6 +218,53 @@ in the file).
 For examples of fields that this construct is recommended for, see the [section of STAC Best Practices](../best-practices.md#common-use-cases-of-additional-fields-for-assets)
 that talks about common use cases of additional fields for assets.
 
+## Media Types
+
+One of the best ways to help inform web clients about the content in a link is to use a common [Media 
+Type](https://en.wikipedia.org/wiki/Media_type) in the `type` field. In STAC the `type` field is used in both the 
+`[link](#link-object)` and `[asset](asset-object)` objects. It is quite useful for STAC browsers to better determine
```
Comment (m-mohr) [inline item-spec/item-spec.md:225]:
```suggestion
`[link](#link-object)` and `[asset](#asset-object)` objects. It is quite useful for STAC browsers to better determine
```

File: item-spec/item-spec.md [hunk 3]
```diff
@@ -252,6 +218,53 @@ in the file).
 For examples of fields that this construct is recommended for, see the [section of STAC Best Practices](../best-practices.md#common-use-cases-of-additional-fields-for-assets)
 that talks about common use cases of additional fields for assets.
 
+## Media Types
```
Comment (m-mohr) [inline item-spec/item-spec.md:221]: Wasn't there an issue to move this to a separate page anyway? This would be a good chance to move it...

Comment (cholmes) [reply item-spec/item-spec.md:221]: There was - #930 I closed it since there was no real consolidation to do, and I don't love specs where you have to keep looking at other pages...

Comment (m-mohr) [reply item-spec/item-spec.md:221]: I still think it would make sense as it's a central part of all three specs and not really bound to items as you've realized yourself. Same for the rel types. At the moment you have to look at other pages anyway for Catalog and Collection, so it's only different for items.

Comment (cholmes) [reply item-spec/item-spec.md:221]: Ok, could you edit it in the way you want? Feel free to make commits on this branch directly. I wasn't sure where/how to make a shared page between the specs, but it seems like you have a way you want it done.

Comment (m-mohr) [reply item-spec/item-spec.md:221]: No clear idea yet to be honest. Could be similar to common metadata? Or should me create for everything common (rel types, media types, common metadata)?

Comment (cholmes) [reply item-spec/item-spec.md:221]: Ok, just made the changes as discussed on the call. Moved most of the table to 'best practices', and kept just our custom (stac and cog) ones in this table.

**Comment by cholmes** [comment]
Ok, just pushed some more changes, to put the list of types in best practices. I think this one is ready for final review now.

File: best-practices.md [hunk 4]
```diff
@@ -156,6 +157,38 @@ than the overall best resolution.
 - `sar:polarizations` ([sar extension](extensions/sar/)): Provide the polarization content and ordering of a specific asset, similar to `eo:bands`.
 - `sar:product_type` ([sar extension](extensions/sar/)): If mixing multiple product types within a single Item, this can be used to specify the product_type for each asset.
 
+## Working with Media Types
```
Comment (m-mohr) [inline best-practices.md:160]: The headings are a bit strange. The h2 has no content and the h3 is the only header in the h2. This is discouraged in writing...

File: item-spec/item-spec.md [hunk 5]
```diff
@@ -252,6 +218,48 @@ in the file).
...
+
+One of the best ways to help inform web clients about the content in a link is to use a common [Media 
+Type](https://en.wikipedia.org/wiki/Media_type) in the `type` field. In STAC the `type` field is used in both the 
+`[Link](#link-object)` and `[Asset](#asset-object)` Objects. It is quite useful for STAC browsers to better determine
+what to render and display to users searching and browsing the catalog.  Media types are often referred to by the 
+now-deprecated term "MIME types".
```
Comment (m-mohr) [inline item-spec/item-spec.md:227]: Not sure, is that a typo?
```suggestion
now deprecated term "MIME types". 
```

File: item-spec/item-spec.md [hunk 6]
```diff
@@ -252,6 +218,48 @@ in the file).
...
+| `application/json; profile=stac-collection`             | A STAC [Collection](../collection-spec/README.md)            |
+| `image/tiff; application=geotiff`                       | GeoTIFF with standardized georeferencing metadata            |
+| `image/tiff; application=geotiff; profile=cloud-optimized` | [Cloud Optimized GeoTIFF](https://www.cogeo.org/) (unofficial). Once there is an [official media type](http://osgeo-org.1560.x6.nabble.com/Media-type-tc5411498.html) it will be added and the proprietary media type here will be deprecated. |
+
+Deprecation notice: GeoTiff previously used the media type `image/vnd.stac.geotiff` and
+Cloud Optimized GeoTiffs used `image/vnd.stac.geotiff; profile=cloud-optimized`.
+Both can still appear in old catalogues, but are deprecated and should be replaced. This may also shift in the future as
+[OGC sorts out the media types](https://github.com/opengeospatial/geotiff/issues/34).
+
+### Media Type for STAC Item
```
Comment (m-mohr) [inline item-spec/item-spec.md:256]: This chapter seems superfluous with the table above.

File: item-spec/item-spec.md [hunk 7]
```diff
@@ -252,6 +218,48 @@ in the file).
...
+would be appropriate; if it is an XML, then `text/xml` is appropriate. For more information on media types as well as a 
+list of [common media types](../best-practices.md#common-media-types-in-stac) used in STAC see the [best practice on 
+working with media types](../best-practices.md#working-with-media-types).
+
+### STAC Media Types
+
+The following table lists the Media Types for STAC, as well as the way [GeoTIFF](https://en.wikipedia.org/wiki/GeoTIFF)'s 
+(and [COG](https://www.cogeo.org)'s) are referenced in STAC.
+
+| Media Type                                              | Description                                                  |
```
Comment (m-mohr) [inline item-spec/item-spec.md:243]: It's a bit strange that the Item Spec has the media types for all "specs", but Collections and Catalogs, where you'd need them first in the hierarchy, have no such table. So I don't think this is the right place to put such a table. This is better suited in catalog actually as there you link to items and catalogs (and collection inherits from there). In Items you don't link to items and catalogs at all.

File: item-spec/item-spec.md (hunk 5 above)
Comment (kylebarron) [reply item-spec/item-spec.md:227]: I don't think it's a typo because both hyphenated words are describing `term`, but I wasn't an English major 😄

Comment (m-mohr) [reply item-spec/item-spec.md:227]: I'm not even a native speaker, just haven't seen it yet with a hypen so was confused.

Comment (kylebarron) [reply item-spec/item-spec.md:227]: Well I was trying to say that even native speakers have a hard time knowing when to use hyphens correctly 😄

File: best-practices.md (hunk 4 above)
Comment (cholmes) [reply best-practices.md:160]: Ah, good catch. I meant to put in more information there on actual practices around using media types, but looks like I forgot to come back and do that. Will work on it now.

File: item-spec/item-spec.md (hunk 5 above)
Comment (cholmes) [reply item-spec/item-spec.md]: I think I didn't write that originally (copied it from another part of the spec). It may technically be correct but I think it's weird, so let's go with a space. Stated in another way - if I actually wrote that it's a typo, as I personally wouldn't mean to write it that way, but perhaps some english major did ;)

File: item-spec/item-spec.md (hunk 7 above)
Comment (cholmes) [reply item-spec/item-spec.md:243]: Good points. I'm not even sure why I did it in items. I agree Catalog seems to be the most appropriate spot for it. Will move it.

Comment (cholmes) [reply item-spec/item-spec.md:243]: > Finally, if there's really just GeoTiff/COG left, also move it to the best practice. It seems strange to just have them listed here in its own chapter.

Yeah, I think I was compelled originally by the idea that 'we're doing our own thing', and it's worth calling out. But I agree it feels a bit weird like this. I'll put it in the best practice. Could be nice to 'set' the mime type somewhere for COG, like in the 'cog spec' or on cogeo.org.

**Comment by cholmes** [comment]
Thanks for the review, I took all the suggestions and tried to make them. It ended up being a lot of moving parts, so another read through is appreciated in case I mixed something up.

**Comment by emmanuelmathot** [comment]
It does not worth an official review but I took the time read all the additions to detail the media types in STAC. I find all the explanations about mime-types very descriptive and complete. The table with the list of the most common types is an excellent initiative as reference for STAC. Especially for general users that do not want to end up in endless discussion thread (cf. https://github.com/opengeospatial/geotiff/issues/34), best way to have a de facto standard ;-). Finally, having the profiled types for STAC helps a lot in the tool development. Thank you for the effort.

File: best-practices.md [hunk 8]
```diff
@@ -156,6 +157,52 @@ than the overall best resolution.
...
+
+### Common Media Types in STAC
+
+The following table lists a number of commonly used media types in STAC. The first two (GeoTIFF and COG) are not fully standardized 
+yet, but reflect the community consensus direction. There are many IANA registered types that commonly show up in STAC. The 
+following table lists some of the most common ones you may encounter or use.
+
+| Media Type                                              | Description                                                  |
+| ------------------------------------------------------- | ------------------------------------------------------------ |
+| `image/tiff; application=geotiff`                       | GeoTIFF with standardized georeferencing metadata            |
```
Comment (m-mohr) [inline best-practices.md:176]: This may change in the future too, but I guess we can't do anything than to stick to it for now...

File: item-spec/item-spec.md [hunk 9]
```diff
@@ -169,12 +169,23 @@ or streamed. It is allowed to add additional fields.
...
+| type        | string    | [Media type](../catalog-spec/catalog-spec.md#media-types) of the asset. See the [common media types](../best-practices.md#common-media-types-in-stac) in the best practice doc for commonly used asset types. |
 | roles       | \[string] | The [semantic roles](#asset-role-types) of the asset, similar to the use of `rel` in links. |
 
 Beyond to the fields above, any of the [additional fields](#additional-fields) *may* be added to the assets. But this
 is recommended only in special cases, see [Additional Fields for Assets](#additional-fields-for-assets)) for more information.
 
+#### Asset Media Type
+
+Any media type can be used in an Item's asset `type` field, and [registered](https://www.iana.org/assignments/media-types/media-types.xhtml) 
+Media Types are preferred. In cases where custom vendor-specific media types are necessary, they should
```
Comment (m-mohr) [inline item-spec/item-spec.md:181]: The custom media types part seems to duplicate the best practice chapter "Formats with no registered media type". I would think it's enough to link there. People usually just use what is available and dont think about new media types or eben registering them. Seems out of scope for the item spec.

**Review by m-mohr — CHANGES_REQUESTED** [review]
Overall looks very good. I made some additional comment, but none of them really blocking the merge.

What is really blocking this is a recent comment from the OGC: https://github.com/opengeospatial/geotiff/issues/34#issuecomment-770897766

I think with that in mind we can't use the proposed media types and need to come up with our own, which are not adaptions from application/json and application/geo+json. Having that said, I'm afraid that we break tooling when servers will return new media types that are unknown. Like currently many HTTP libraries will automatically parse JSON due to the media type, but once servers return application/stac-item or so, that will probably not happen any more. Even browsers may offer you stac files for download then instead of just serving them nicely formatted as JSON. I think this may lead to more issues than it actually solves and thus my preference at the moment would be to stick with the JSON and GeoJSON media types. Maybe things impove a bit with the `type` field we add to Catalogs and Collections in #889. If we decide to have new media types, we should run intensive tests beforehand with a number of common implementations. This would delay RC1 by a lot, but on the same hand we should probably have it in for the RC phase.

**Comment by cholmes** [comment]
Yeah, I agree that we should stick with json / geojson types, and then drop our 'profile' stuff. It seems far more useful for clients to know that the linked files are json / geojson than and be able to do 'something' with it, vs 'application/vnd.stac-catalog' or whatever that most clients won't recognize. One idea would be to add some extra parameter to our 'link' object, so you'd say 'type=application/json' and then 'type-profile' or 'sub-type' or something equals 'stac-catalog'.

**Comment by cholmes** [comment]
As discussed in #851 and on a call - our proposed rel types don't really work with IANA. So we must pick between a custom type that we register, which would mean applications that understand json or geojson mime types would ignore ours, or not having 'our own' mime type. The consensus was that having our own doesn't add enough value to overcome things like a browser no longer parsing and displaying STAC files (since they wouldn't be json). #944 

This PR did have some nice refactoring, but I'm going to close the issue but keep the branch open and make a new PR when it's in good shape.
